buildSrc/src/main/groovy/org/elasticsearch/gradle/RestTestsFromSnippetsTask.groovy

```diff
@@ -0,0 +1,159 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.gradle
+
+import org.gradle.api.InvalidUserDataException
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.OutputDirectory
+
+import java.util.regex.Matcher
+
+/**
+ * Generates REST tests for each snippet marked // TEST.
+ */
+public class RestTestsFromSnippetsTask extends SnippetsTask {
+
+    @Input
+    String docRoot = project.file('.') // There should be a better way!
+
+    @OutputDirectory
+    File outputRoot = project.file('build/rest')
+
+    /**
+     * The last // TEST we've seen. Implementation detail bu public because
+     * groovy.
+     */
+    Snippet lastTest
+    /**
+     * The last // TESTRESPONSE we've seen. Implementation detail but public
+     * because groovy.
+     */
+    Snippet lastResponse
+
+    /**
+     * The file in which we saw the last test.
+     */
+    private File lastTestFile
+    /**
+     * The file we're building. Implementation detail but public because groovy.
+     */
+    PrintWriter underConstruction
+
+    public RestTestsFromSnippetsTask() {
+        // TODO I bet we can skip files that haven't changed
```

Comment:
This'd be nice but I don't think it is important enough to block the PR - the whole operation takes a couple seconds and it already is skipped if the doc files don't change. So repeating it for all doc files if any change is OK - not great, but ok.